### PR TITLE
src: Add on prem blueprints import support (HMS-4683)

### PR DESCRIPTION
**Pull request by avitova**
This pull request relates to [#4683](https://issues.redhat.com/browse/HMS-4683).
We want to support import on prem blueprints in our import feature. This enables importing toml blueprints that get automatically detected, and parsed.
A lot of fields we need to ignore, a lot of fields from our blueprints are not in the on-prem blueprint structure. This is our first attempt to allow on-prem blueprint, and it will get us more information whether users would use it or not. We also need to add a method to measure the usage of this feature. This is not here yet, and I would like to do it as a follow up, as it can raise discussions, and we do not want to block this feature on it.

We will bring a different solution, so eventually, we will get rid of this mapper. But the other solution that should be more "bulletproof" will take time, in the meantime we can use this one. :)

I know, that on-prem blueprints can also come in json, this will be quite easy to do, we will adjust the UI a bit for that, so that users would be able to import even on-prem json.

Feel free to ask any questions, and find bugs!

### CODE REVIEW

**Review by ezr-ondrej — COMMENTED** [review]
What's the status here? should I test it out?

File: src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx [hunk 1]
```diff
@@ -77,7 +77,7 @@ const PackageRecommendations = () => {
           },
         });
 
-        if (response && response.data && response.data.packages.length > 0) {
+        if (response && response.data && response.data.packages && response.data.packages.length > 0) {
```
Comment (avitova) [inline src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx:80]: This is not related, but I was getting some errors, and I believe that we need to check for this too.

File: src/Components/Blueprints/ImportBlueprintModal.tsx [hunk 2]
```diff
@@ -50,52 +53,63 @@ export const ImportBlueprintModal: React.FunctionComponent<
...
+        const isToml = filename.endsWith('.toml');
+        const isJson = filename.endsWith('.json');
+        if (isToml) {
+          const tomlBlueprint = parse(fileContent);
+          const blueprintFromFile = mapOnPremToHosted(tomlBlueprint);
+          const importBlueprintState = mapExportRequestToState(
+            blueprintFromFile,
+            []
+          );
+          setImportedBlueprint(importBlueprintState);
```
Comment (ezr-ondrej) [inline src/Components/Blueprints/ImportBlueprintModal.tsx:73]: Should we show some warning, that this is a best effort import?

File: src/Components/Blueprints/ImportBlueprintModal.tsx [hunk 3]
```diff
@@ -50,52 +53,63 @@ export const ImportBlueprintModal: React.FunctionComponent<
...
+            blueprintFromFile.image_requests || []
+          );
+          setImportedBlueprint(importBlueprintState);
+        }
+      } catch (error) {
+        setIsInvalidFormat(true);
+        dispatch(
+          addNotification({
+            variant: 'warning',
+            title: 'No blueprint was build',
```
Comment (ezr-ondrej) [inline src/Components/Blueprints/ImportBlueprintModal.tsx:94]: I know this is just a copy, but the warning is probably incorrect, we should say something like "File is not a valid blueprint"

Comment (ezr-ondrej) [reply src/Components/Blueprints/ImportBlueprintModal.tsx]: Mocks are not always 100% tought through, this case I feel like it could have been copy pasted from other mock.
What I'm interested in is whether my comment makes sense to you, or if I just don't get why this might be actually correct message :)

**Comment by avitova** [comment]
Okay, I addressed all the comments, thanks for the review!
/retest

File: src/Components/Blueprints/ImportBlueprintModal.tsx (hunk 3 above)
Comment (avitova) [reply src/Components/Blueprints/ImportBlueprintModal.tsx]: Well, I feel like the notification might make sense. We already inform in the warning at the bottom of the text field that the format is not compatible with our blueprints. So this one just made sense to me, so that users know that we did not create any blueprint just from the fact that they uploaded the file. We can discuss this with Kelsea, but I think this is a minor thing:)

File: src/Components/Blueprints/ImportBlueprintModal.tsx (hunk 2 above)
Comment (avitova) [reply src/Components/Blueprints/ImportBlueprintModal.tsx:75]: Let me know how you like that one I came up with. We can adjust it. :)

File: src/Components/Blueprints/ImportBlueprintModal.tsx (hunk 3 above)
Comment (ezr-ondrej) [reply src/Components/Blueprints/ImportBlueprintModal.tsx]: Yeah, totaly minor, if you feel this is good, lets keep it :)

File: src/Components/Blueprints/ImportBlueprintModal.tsx [hunk 4]
```diff
@@ -33,69 +36,83 @@ export const ImportBlueprintModal: React.FunctionComponent<
...
   };
-  const [jsonContent, setJsonContent] = React.useState('');
+  const [fileContent, setFileContent] = React.useState('');
   const [importedBlueprint, setImportedBlueprint] =
     React.useState<wizardState>();
   const [isInvalidFormat, setIsInvalidFormat] = React.useState(false);
   const [filename, setFilename] = React.useState('');
   const [isLoading, setIsLoading] = React.useState(false);
   const [isRejected, setIsRejected] = React.useState(false);
+  const [isOnPrem, setIsOnPrem] = React.useState(false);
```
Comment (ezr-ondrej) [inline src/Components/Blueprints/ImportBlueprintModal.tsx:50]: You might want to reset this flag in `onClear` handler.

**Review by ezr-ondrej — APPROVED** [review]
I'm good here, but merging with the current CI state might be difficult xD

FTR: I've tested this with toml and seems fine, just one nit :)

File: src/Components/Blueprints/ImportBlueprintModal.tsx (hunk 4 above)
Comment (avitova) [reply src/Components/Blueprints/ImportBlueprintModal.tsx:51]: Good catch, thank you! I added more resets to the default value, you are right.

**Review by ezr-ondrej — DISMISSED** [review]
Thanks, looks great! :+1:

File: src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx [hunk 5]
```diff
@@ -77,7 +77,12 @@ const PackageRecommendations = () => {
           },
         });
 
-        if (response && response.data && response.data.packages.length > 0) {
+        if (
```
Comment (mgold1234) [inline src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx:80]: is that change related to the pr?

**Review by mgold1234 — COMMENTED** [review]
this support of import on Prem is amazing!

File: src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx (hunk 5 above)
Comment (avitova) [reply src/Components/CreateImageWizard/steps/Packages/PackageRecommendations.tsx:80]: This is not related. However, we access the `response.data.packages`. This was a bug, that probably was not tested, but with import, I found out that this was failing. And I think we should not really use methods that are not safe for empty arrays before checking `response.data.packages`.

**Review by regexowl — COMMENTED** [review]
This looks great, thank you! 

Found one problem, the Wizard crashes for me when I get to the packages step because of the PackageRecommendations, adding a screenshot of the error:

image

Tried to temporarily comment out PackageRecommendations to see other imported values and everything else looks amazing :heart:

**Comment by regexowl** [comment]
I think adding a new format to the import options shouldn't have effect on existing IQE tests, but not completely sure - what do you think @jrusz?

**Comment by regexowl** [comment]
Ah, this seems to help with the crashes:
image

`PackageRecommendations.tsx`, line 112

**Comment by regexowl** [comment]
Will address the package recommendations boo boos in https://github.com/osbuild/image-builder-frontend/pull/2606

**Comment by jrusz** [comment]
> I think adding a new format to the import options shouldn't have effect on existing IQE tests, but not completely sure - what do you think @jrusz?

Yeah it should be fine as long as the old workflow and format works.

**Comment by ezr-ondrej** [comment]
@avitova now there are merge conflicts with the fix, can you please rebase, it seems apart of that we are ready :)

File: src/Components/Blueprints/ImportBlueprintModal.test.tsx [hunk 6]
```diff
@@ -207,4 +260,79 @@ describe('Import modal', () => {
...
+    await screen.findByRole('textbox', {
+      name: 'Select activation key',
+    });
+    await clickNext();
+
+    // OpenScap
+    await clickNext();
+
+    //File system configuration
+    await clickNext();
```
[the remark this answers is not in the document]
Comment (avitova) [reply src/Components/Blueprints/ImportBlueprintModal.test.tsx:324]: Yeah, we might. :thinking: And I think it could show to us more, what values we are able to import from on prem. :) But would you be okay if we just did that in a followup maybe? :thinking:

**Comment by avitova** [comment]
I rebased and pinned the version. Thank you for your reviews. :)

**Review by regexowl — APPROVED** [review]
Looks great, thank you!
